cpp/src/join/join.cu

```diff
@@ -34,10 +35,15 @@ inner_join(table_view const& left_input,
            rmm::cuda_stream_view stream,
            rmm::mr::device_memory_resource* mr)
 {
+  // flatten any structs out. Note this happens before dictionary matching because
+  // structs can contain dictionaries.
+  auto const flattened_left  = structs::detail::flatten_nested_columns(left_input, {}, {});
+  auto const flattened_right = structs::detail::flatten_nested_columns(right_input, {}, {});
```

Comment:
I think this should just happen in the `hash_join` ctor and `probe`. That way these calls don't need to be repeated.

Reply:
I originally had it this way, but I was worried that the dictionary matching wouldn't happen for a dictionary inside a struct. The dictionary code is duplicated as well. Should we move both inside `hash_join` and `probe`?

Reply:
The dictionary code can't be moved inside the `hash_join` as it requires knowing both of the dictionary columns at the same time.

Reply:
Spoke with Jake and we don't need to run the flatten before the dictionary work, so I have moved this code into the `hash_join` object as requested. This reduces the number places for this flattening code to a single class.